src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -25,33 +17,41 @@
 
 @WebServlet("/login")
 public class LoginServlet extends HttpServlet {
+  private String clientId =
+      "12523748853-6h3fnh1kppht9s895ist2b5222ifh2u7.apps.googleusercontent.com"; // Client ID.
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    UserService userService = UserServiceFactory.getUserService();
-    boolean loggedIn = userService.isUserLoggedIn();
-    String urlToRedirect = "/search.html";
-
-    JSONObject userDetails = new JSONObject();
-    userDetails.put("loggedIn", loggedIn);
-    if (loggedIn) {
-      String userEmail = userService.getCurrentUser().getEmail();
-      userDetails.put("userEmail", userEmail);
-      String logoutURL = userService.createLogoutURL(urlToRedirect);
-      userDetails.put("logoutURL", logoutURL);
-      // TODO: Implement OAuth for more robust email checking.
-      int start = userEmail.indexOf('@');
-      int end = userEmail.lastIndexOf('.');
-      if (start != -1 && end != -1) {
-        String schoolName = userEmail.substring(start + 1, end);
-        userDetails.put("schoolName", schoolName);
+  public LoginObject postHelper(HttpServletRequest request) throws IOException {
+    String idTokenString = request.getParameter("token");
+    System.out.println(idTokenString);
+    String id = "NOT FOUND";
+    Boolean success = false;
+    UrlFetchTransport transport = UrlFetchTransport.getDefaultInstance();
+    GsonFactory json = GsonFactory.getDefaultInstance();
+    GoogleIdTokenVerifier verifier =
+        new GoogleIdTokenVerifier.Builder(transport, json)
+            .setAudience(Collections.singletonList(clientId))
+            .build();
+    try {
+      GoogleIdToken idToken = verifier.verify(idTokenString);
+      if (idToken != null) {
+        Payload payload = idToken.getPayload();
+        String userId = payload.getSubject();
+        id = userId;
+        success = true;
       }
-
-    } else {
-      String loginURL = userService.createLoginURL(urlToRedirect);
-      userDetails.put("loginURL", loginURL);
+      return new LoginObject(id, success);
+    } catch (GeneralSecurityException | IllegalArgumentException e) {
+      return new LoginObject(id, success);
```

Comment:
Not sure if it is really relevant but in this case would you error LoginObject not look exactly like a case where there is no user? Rather than have them identical the error one could have and id= "Error" or something to let you know it came from an error

Reply:
I actually implemented this, but then talked to @Tetheta and we decided to remove the Test servlet entirely as the Library doesn't really support it (link [here](https://github.com/googleapis/google-api-java-client/issues/973))